File: LeetCodeKotlin/BalancedBinaryTree_110.kt
class BalancedBinaryTree_110 {
    fun isBalanced(root: TreeNode?): Boolean {
        return checkHeight(root) != -1
    }

    private fun checkHeight(node: TreeNode?): Int {
        if (node == null) return 0

        val leftHeight = checkHeight(node.left)
        if (leftHeight == -1) return -1 // Left subtree not balanced

        val rightHeight = checkHeight(node.right)
        if (rightHeight == -1) return -1 // Right subtree not balanced

        if (kotlin.math.abs(leftHeight - rightHeight) > 1) {
            return -1 // Current node not balanced
        }

        return maxOf(leftHeight, rightHeight) + 1
    }
}

File: LeetCodeCpp/problems/trees/BalancedBinaryTree_110.cpp
#include "BalancedBinaryTree_110.h"

#include <algorithm>
#include <cstdlib>

using namespace std;

bool BalancedBinaryTree_110::isBalanced(TreeNode<int>* root) {
    return heightOrUnbalanced(root) != -1;
}

int BalancedBinaryTree_110::heightOrUnbalanced(TreeNode<int>* node) {
    if (node == nullptr) {
        return 0;
    }

    const int leftHeight = heightOrUnbalanced(node->left);
    if (leftHeight == -1) {
        return -1;
    }

    const int rightHeight = heightOrUnbalanced(node->right);
    if (rightHeight == -1) {
        return -1;
    }

    if (abs(leftHeight - rightHeight) > 1) {
        return -1;
    }

    return 1 + max(leftHeight, rightHeight);
}
